libevm/lib/service_state.go

```diff
@@ -0,0 +1,249 @@
+	return nil
+}
+
+func (s *Service) StateSubBalance(params BalanceParams) error {
+	err, statedb := s.getState(params.Handle)
+	if err != nil {
+		return err
+	}
+	statedb.SubBalance(params.Address, params.Amount.ToInt())
+	return nil
+}
+
+func (s *Service) StateSetBalance(params BalanceParams) error {
+	err, statedb := s.getState(params.Handle)
+	if err != nil {
+		return err
+	}
+	statedb.SetBalance(params.Address, params.Amount.ToInt())
+	return nil
+}
+
+func (s *Service) StateGetNonce(params AccountParams) (error, uint64) {
+	err, statedb := s.getState(params.Handle)
+	if err != nil {
+		return err, 0
+	}
+	return nil, statedb.GetNonce(params.Address)
+}
+
+func (s *Service) StateSetNonce(params NonceParams) error {
+	err, statedb := s.getState(params.Handle)
+	if err != nil {
+		return err
+	}
+	statedb.SetNonce(params.Address, params.Nonce)
+	return nil
+}
+
+func (s *Service) StateGetCodeHash(params AccountParams) (error, common.Hash) {
+	err, statedb := s.getState(params.Handle)
+	if err != nil {
+		return err, common.Hash{}
+	}
+	return nil, statedb.GetCodeHash(params.Address)
+}
+
+// make sure that the given account is not declared "empty" and then removed
+func preserveAccount(statedb *state.StateDB, addr common.Address) {
+	obj := statedb.GetOrNewStateObject(addr)
+	// this is basically a copy of stateObject.empty() which is not exported and therefore not accessible here
+	empty := obj.Nonce() == 0 && obj.Balance().Sign() == 0 && bytes.Equal(obj.CodeHash(), emptyCodeHash)
+	if empty {
+		obj.SetCode(crypto.Keccak256Hash(addr.Bytes()), make([]byte, 0))
+	}
+}
+
+var chunkedMagicValue = common.HexToHash("0x55253b59b75c20e73965d68d3673a76f42f6415752c6fa3ce7a11c3ee5689aec")
+
+// chunk keys are generating by hashing the original key and the chunk index
+func getChunkKey(key common.Hash, chunkIndex int) common.Hash {
+	chunkIndexBytes := big.NewInt(int64(chunkIndex)).Bytes()
+	hashedData := append(key.Bytes(), chunkIndexBytes...)
+	return crypto.Keccak256Hash(hashedData)
+}
+
+func skipLeadingZeroes(data common.Hash) []byte {
+	data.Bytes()
+	for i, b := range data.Bytes() {
+		if b != 0 {
+			return data.Bytes()[i:]
+		}
+	}
+	return nil
+}
+
+func (s *Service) StateGetStorage(params StorageParams) (error, []byte) {
+	err, statedb := s.getState(params.Handle)
+	if err != nil {
+		return err, nil
+	}
+	value := statedb.GetState(params.Address, params.Key)
+	if value != chunkedMagicValue {
+		return nil, skipLeadingZeroes(value)
+	}
+	length := int(statedb.GetState(params.Address, getChunkKey(params.Key, 0)).Big().Int64())
+	data := make([]byte, length)
+	for i := 0; i < length; i += common.HashLength {
+		end := i + common.HashLength
+		if end > length {
+			end = length
+		}
+		chunk := statedb.GetState(params.Address, getChunkKey(params.Key, i+1))
+		copy(data[i:end], skipLeadingZeroes(chunk))
+	}
+	return nil, data
+}
+
+// StateSetStorage writes values of arbitrary length to the storage trie of given account
```

Comment:
The algorithm for storing values larger than 32 bytes works but is not particularly pretty.